dashboard/app/views/devise/registrations/new.html.haml

```diff
@@ -119,22 +123,44 @@
             %hr
             %h1.custom-h1= t('terms_interstitial.privacy_notice')
             = render partial: 'home/privacy_notice'
-          %br
-          #terms_of_service_version-block.itemblock
-            .error_in_field.checkbox-error
-            %br
-            = f.label :terms_of_service_version do
-              = f.check_box :terms_of_service_version, {class: 'terms-checkbox', :checked => true}, User::TERMS_OF_SERVICE_VERSIONS.last
-              = t('terms_interstitial.accept_label', tos_url: CDO.code_org_url('/tos'), privacy_url: CDO.code_org_url('/privacy')).html_safe
 
+        -# Show terms of service for both student and teacher
         %br
+        #terms_of_service_version-block.itemblock
+          .error_in_field.checkbox-error
+          %br
+          = f.label :terms_of_service_version do
+            = f.check_box :terms_of_service_version, {class: 'terms-checkbox'}, User::TERMS_OF_SERVICE_VERSIONS.last
+            -# Teacher text.
+            %span.teacher-options{style: 'display: none'}
+              = t('terms_interstitial.accept_label', tos_url: CDO.code_org_url('/tos'), privacy_url: CDO.code_org_url('/privacy')).html_safe
+            -# Student text.
+            %span.student-options
+              != t('signup_form.agree_tos_privacy')
+              != t('signup_form.under_13_consent')
+
+        -# If GDPR applies, show an additional checkbox.
+        #data_transfer_agreement_accepted-block.itemblock
+          - if request.gdpr? || request.params['force_in_eu']
+            .error_in_field.checkbox-error
+            = f.hidden_field :data_transfer_agreement_required, value: "1"
```

Comment:
@breville I think this might be a mistake.  All of our validation around the data transfer agreement fields hinges on this `"1"` being sent along with the request, but by making it a hidden field on the form we are trusting the client. A bad actor could update the value of this hidden field before submitting the form and bypass our validation.

Is there a reason we can't merge this virtual attribute into the request parameters in the controller, taking it out of the client's control?

Alternatively, do we need this attribute at all?  Validation starts with this check

https://github.com/code-dot-org/code-dot-org/blob/8f719b8aeb0a87640da3670e3b785a6466af12e5/dashboard/app/models/user.rb#L523

[The docs for the `validates :x, acceptance: true` check](http://guides.rubyonrails.org/active_record_validations.html#acceptance) _claim_ it only runs if `x` is not nil, which would make the conditional clause in our validation moot.  This test passes...

https://github.com/code-dot-org/code-dot-org/blob/8f719b8aeb0a87640da3670e3b785a6466af12e5/dashboard/test/controllers/registrations_controller_test.rb#L202-L210

...because the controller is coercing the acceptance parameter to a boolean before letting the model do its validation.

https://github.com/code-dot-org/code-dot-org/blob/f83f025af8312196de52a25ae0baa515a1f88549/dashboard/app/controllers/registrations_controller.rb#L59

All that to say, I think we're jumping through extra hoops here _and_ possibly letting users sign up without agreeing.  I also noticed Erin's current work on the GDPR dialog doesn't send along this extra parameter, so validation doesn't run.  How do you feel about attempting some cleanup together?

Reply:
I thought about relying on the incoming form submission's request, but decided to rely on a hidden field in the form because a component of the GDPR compliance is having an audit trail in which we know what we presented to the user.  In the unlikely case that we presented them a UI with EU-specific fields, but then their submission didn't appear to come from the same IP address, then we might have a mismatch between what what we presented (and what they accepted) and what we then attempt to record.

If a bad actor was messing with these fields then that could also affect other important fields in sign-up, couldn't it, like ToS acceptance?  I'm not sure how likely a bad actor would be in this situation, but maybe I'm missing an obvious scenario?

Reply:
I think you're right that the model should ensure that `data_transfer_agreement_accepted` is `true`, not just non-`nil`, when `data_transfer_agreement_required` is `true`.

(Incidentally, we coerce the values to be boolean rather than "0" or "1" in the controller is due to the last line here: http://guides.rubyonrails.org/active_record_validations.html#acceptance.)

Reply:
This last line?

> If the field does exist in your database, the accept option must be set to or include `true` or else the validation will not run.

The default value of the `:accept` option (which we're not providing) is `['1', true]`, so it already includes `true`.

Or a different last line?

Reply:
Oh, I had misinterpreted that to mean we had to coerce the option itself.  So I guess the coercion wasn't necessary after all.

Reply:
That said, I did like providing the model with booleans rather than `"1"`s anyway...

Reply:
We chatted more about this offline, and are leaving things as-is for now.  The `data_transfer_agreement_required` and `data_transfer_agreement_accepted` are deliberately coerced to booleans in the registrations controller so that the model can deal with them as explicitly true and false, and when `data_transfer_agreement_accepted` is written as a serialized attribute it ensures it's a boolean and not a string.  The pattern to have a `data_transfer_agreement_required` attribute is inspired by what we did previously for `email_preference_opt_in_required` and is a neat trick to ensure that the model's validations enforce the required field when it's, well, required.  The reason we want the model's validations to check for this sometimes-required field is somewhat unique to the sign-up UI, because it uses the errors sent back from the model to the client to display *all* missing or invalid fields to the user at the same time.